README.md

## ewenthemes

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![](http://www.r-pkg.org/badges/version/ewenthemes)](http://www.r-pkg.org/pkg/ewenthemes)
[![Travis-CI Build
Status](https://travis-ci.org/ewenme/ewenthemes.svg?branch=master)](https://travis-ci.org/ewenme/ewenthemes)

Minimal, typography-centric themes and aesthetic helpers, principally
for charts on [ewen.io](https://ewen.io). Standing on the shoulders of
[hrbrthemes](https://github.com/hrbrmstr/hrbrthemes).

The core theme, `theme_ewen()`, uses Arial Narrow, which should be
installed on practically any modern system. Variations on this theme
implement custom typefaces - `theme_ewen_rs` uses the [Recursive
typography palette](https://www.recursive.design/), `theme_ewen_ws()`
runs on Wei Huang’s [Work
Sans](https://github.com/weiweihuanghuang/Work-Sans).

Core themes:

  - `theme_ewen`: Arial Narrow-based theme
  - `theme_ewen_rs`: Recursive Sans-based theme
  - `theme_ewen_ws`: Work Sans-based theme

### Installation

``` r
devtools::install_github("ewenme/ewenthemes")
```

### Usage

``` r
library(ewenthemes)
library(ggplot2)
```

### Base theme + Recursive Sans

``` r
ggplot(mtcars, aes(mpg, wt)) +
  geom_point() +
  labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
       title="Seminal ggplot2 scatterplot example",
       subtitle="A plot that is only useful for demonstration purposes",
       caption="Brought to you by the letter 'g'") + 
  # apply theme
  theme_ewen_rs()
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="672" />

### Base theme + Work Sans

``` r
ggplot(mtcars, aes(mpg, wt)) +
  geom_point() +
  labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
       title="Seminal ggplot2 scatterplot example",
       subtitle="A plot that is only useful for demonstration purposes",
       caption="Brought to you by the letter 'g'") + 
  # apply theme
  theme_ewen_ws()
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="672" />


NEWS.md
# ewenthemes 1.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/test-all.R
library(testthat)
test_check("ewenthemes")


tests/testthat/test-ewenthemes.R
context("core theme components work")
test_that("we can do something", {

  th <- theme_ewen(grid="XY", axis="xy")
  expect_that(th$plot.title$family, equals("Arial Narrow"))

  thrc <- theme_ewen_ws(grid="XY", axis="xy")
  expect_that(thrc$plot.title$family, equals("Work Sans"))

  invisible(theme_ewen(grid=FALSE))
  invisible(theme_ewen_ws(grid=FALSE))
  invisible(theme_ewen(grid="XY"))
  invisible(theme_ewen_ws(grid="XY"))
  invisible(theme_ewen(grid="xy"))
  invisible(theme_ewen_ws(grid="xy"))

  invisible(theme_ewen(axis=TRUE))
  invisible(theme_ewen_ws(axis=TRUE))

  invisible(theme_ewen(axis=FALSE))
  invisible(theme_ewen_ws(axis=FALSE))
  invisible(theme_ewen(axis="xy"))
  invisible(theme_ewen_ws(axis="xy"))
  invisible(theme_ewen(axis=""))
  invisible(theme_ewen_ws(axis=""))

  testthat::skip_on_cran()
  invisible(import_work_sans())

})


R/spectral.R
#' A precise & pristine [ggplot2] theme with an emphasis on typography,
#' supporting the \href{https://ewen.io/}{ewen.io} site, based on
#' `theme_ipsum` from \href{https://github.com/hrbrmstr/hrbrthemes}{hrbrthemes}.
#'
#' You should [import_spectral]() first and also install the fonts on your
#' system before trying to use this theme.
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @md
#' @param base_family,base_size base font family and size
#' @param plot_title_family,plot_title_face,plot_title_size,plot_title_margin plot title family, face, size and margi
#' @param subtitle_family,subtitle_face,subtitle_size plot subtitle family, face and size
#' @param subtitle_margin plot subtitle margin bottom (single numeric value)
#' @param strip_text_family,strip_text_face,strip_text_size facet label font family, face and size
#' @param caption_family,caption_face,caption_size,caption_margin plot caption family, face, size and margin
#' @param axis_title_family,axis_title_face,axis_title_size axis title font family, face and size
#' @param axis_title_just axis title font justification, one of `[blmcrt]`
#' @param plot_margin plot margin (specify with [ggplot2::margin()])
#' @param grid_col,axis_col grid & axis colors; grid defaults to `#cccccc`, axes default to `#2b2b2b`
#' @param grid panel grid (`TRUE`, `FALSE`, or a combination of `X`, `x`, `Y`, `y`)
#' @param axis_text_size font size of axis text
#' @param axis add x or y axes? `TRUE`, `FALSE`, "`xy`"
#' @export
#' @examples \dontrun{
#' library(ggplot2)
#'
#' # seminal scatterplot
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_ewen_sp()
#' }
theme_ewen_sp <- function(base_family="Spectral", base_size = 11.5,
                       plot_title_family=if (.Platform$OS.type == "windows") "Spectral" else "Spectral Bold",
                       plot_title_size = 18,
                       plot_title_face="bold", plot_title_margin = 10,
                       subtitle_family=if (.Platform$OS.type == "windows") "Spectral" else "Spectral Light",
                       subtitle_size = 12,
                       subtitle_face = "plain", subtitle_margin = 15,
                       strip_text_family = base_family, strip_text_size = 12,
                       strip_text_face = "plain",
                       caption_family = if (.Platform$OS.type == "windows") "Spectral" else "Spectral Light",
                       caption_size = 11,
                       caption_face = "plain", caption_margin = 15,
                       axis_text_size = base_size,
                       axis_title_family = subtitle_family,
                       axis_title_size = base_size,
                       axis_title_face = "plain", axis_title_just = "rt",
                       plot_margin = margin(10, 10, 10, 10),
                       grid_col = "#cccccc", grid = "Y",
                       axis_col = "#2b2b2b", axis = "x"
) {

  ret <- ggplot2::theme_minimal(base_family=base_family, base_size=base_size)

  ret <- ret + theme(legend.background=element_blank())
  ret <- ret + theme(legend.key=element_blank())

  if (inherits(grid, "character") | grid == TRUE) {

    ret <- ret + theme(panel.grid=element_line(color=grid_col, size=0.2))
    ret <- ret + theme(panel.grid.major=element_line(color=grid_col, size=0.2))
    ret <- ret + theme(panel.grid.minor=element_line(color=grid_col, size=0.15))

    if (inherits(grid, "character")) {
      if (regexpr("X", grid)[1] < 0) ret <- ret + theme(panel.grid.major.x=element_blank())
      if (regexpr("Y", grid)[1] < 0) ret <- ret + theme(panel.grid.major.y=element_blank())
      if (regexpr("x", grid)[1] < 0) ret <- ret + theme(panel.grid.minor.x=element_blank())
      if (regexpr("y", grid)[1] < 0) ret <- ret + theme(panel.grid.minor.y=element_blank())
    }

  } else {
    ret <- ret + theme(panel.grid=element_blank())
  }

  if (inherits(axis, "character") | axis == TRUE) {
    ret <- ret + theme(axis.line=element_line(color="#2b2b2b", size=0.5))
    if (inherits(axis, "character")) {
      axis <- tolower(axis)
      if (regexpr("x", axis)[1] < 0) {
        ret <- ret + theme(axis.line.x=element_blank())
      } else {
        ret <- ret + theme(axis.line.x=element_line(color=axis_col, size=0.5))
      }
      if (regexpr("y", axis)[1] < 0) {
        ret <- ret + theme(axis.line.y=element_blank())
      } else {
        ret <- ret + theme(axis.line.y=element_line(color=axis_col, size=0.5))
      }
    } else {
      ret <- ret + theme(axis.line.x=element_line(color=axis_col, size=0.5))
      ret <- ret + theme(axis.line.y=element_line(color=axis_col, size=0.5))
    }
  } else {
    ret <- ret + theme(axis.line=element_blank())
  }

  ret <- ret + theme(axis.ticks = element_blank())
  ret <- ret + theme(axis.ticks.x = element_blank())
  ret <- ret + theme(axis.ticks.y = element_blank())

  xj <- switch(tolower(substr(axis_title_just, 1, 1)), b=0, l=0, m=0.5, c=0.5, r=1, t=1)
  yj <- switch(tolower(substr(axis_title_just, 2, 2)), b=0, l=0, m=0.5, c=0.5, r=1, t=1)

  ret <- ret + theme(axis.text.x=element_text(size=axis_text_size, margin=margin(t=0)))
  ret <- ret + theme(axis.text.y=element_text(size=axis_text_size, margin=margin(r=0)))
  ret <- ret + theme(axis.title=element_text(size=axis_title_size, family=axis_title_family))
  ret <- ret + theme(axis.title.x=element_text(hjust=xj, size=axis_title_size,
                                               family=axis_title_family, face=axis_title_face))
  ret <- ret + theme(axis.title.y=element_text(hjust=yj, size=axis_title_size,
                                               family=axis_title_family, face=axis_title_face))
  ret <- ret + theme(axis.title.y.right=element_text(hjust=yj, size=axis_title_size, angle=90,
                                                     family=axis_title_family, face=axis_title_face))
  ret <- ret + theme(strip.text=element_text(hjust=0, size=strip_text_size,
                                             face=strip_text_face, family=strip_text_family))
  ret <- ret + theme(panel.spacing=grid::unit(2, "lines"))
  ret <- ret + theme(plot.title=element_text(hjust=0, size=plot_title_size,
                                             margin=margin(b=plot_title_margin),
                                             family=plot_title_family, face=plot_title_face))
  ret <- ret + theme(plot.subtitle=element_text(hjust=0, size=subtitle_size,
                                                margin=margin(b=subtitle_margin),
                                                family=subtitle_family, face=subtitle_face))
  ret <- ret + theme(plot.caption=element_text(hjust=1, size=caption_size,
                                               margin=margin(t=caption_margin),
                                               family=caption_family, face=caption_face))
  ret <- ret + theme(plot.margin=plot_margin)

  ret

}


#' Import Work Sans font for use in charts
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @md
#' @note This will take care of ensuring PDF/PostScript usage. The location of the
#'   font directory is displayed after the base import is complete. It is highly
#'   recommended that you install them on your system the same way you would any
#'   other font you wish to use in other programs.
#' @export
import_spectral <- function() {

  sp_font_dir <- system.file("fonts", "spectral", package="ewenthemes")

  suppressWarnings(suppressMessages(extrafont::font_import(sp_font_dir, prompt=FALSE)))

  message(
    sprintf(
      "You will likely need to install these fonts on your system as well.\n\nYou can find them in [%s]",
      sp_font_dir)
  )

}

#' @rdname Spectral
#' @md
#' @title Work Sans font name R variable aliases
#' @description `font_sp` == "`Spectral`"
#' @format length 1 character vector
#' @export
font_sp <- "Spectral"

#' @rdname Spectral
#' @md
#' @note `font_sp_bold` (a.k.a. "`Spectral Bold`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_sp_bold` == "`Spectral Bold`"
#' @export
font_sp_bold <- "Spectral Bold"

#' @rdname Spectral
#' @md
#' @note `font_sp_light` (a.k.a. "`Spectral Light`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_sp_light` == "`SpectralLight`"
#' @export
font_sp_light <- "Spectral Light"


R/recursive.R
#' [ggplot2] theme supporting the \href{https://ewen.io/}{ewen.io} site, based on
#' `theme_ipsum` from \href{https://github.com/hrbrmstr/hrbrthemes}{hrbrthemes}.
#'
#' You should [import_recursive]() first and also install the fonts on your
#' system before trying to use this theme.
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @inheritParams theme_ewen
#'
#' @export
#'
#' @examples \dontrun{
#' library(ggplot2)
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_ewen_rs()
#' }

theme_ewen_rs <- function(
  base_family = "Recursive Regular", base_colour = "#111111", base_size = 14,
  base_margin = 10, base_face = "plain",
  plot_title_family = "Recursive Black", plot_title_size = 24,
  plot_title_face = "plain", plot_title_margin = 10,
  subtitle_family = base_family, subtitle_size = base_size,
  subtitle_face = base_face, subtitle_margin = base_margin,
  axis_text_family = base_family, axis_text_size = base_size,
  axis_title_family = base_family, axis_title_size = base_size,
  axis_title_face = base_face, axis_title_margin = base_margin,
  axis_title_just = "rt",
  caption_family = "Recursive Light", caption_size = base_size,
  caption_face = base_face, caption_margin = base_margin*1.5,
  strip_text_family = base_family, strip_text_size = base_size,
  strip_text_face = base_face,
  plot_margin = c(10, 10, 10, 10),
  grid_col = "#cccccc", grid = "Y",
  axis_col = "#9B9B9B", axis = "x"
  ) {

  theme_ewen(
    base_family = base_family, base_colour = base_colour, base_size = base_size,
    base_margin = base_margin, base_face = base_face,
    plot_title_family = plot_title_family, plot_title_size = plot_title_size,
    plot_title_face = plot_title_face, plot_title_margin = plot_title_margin,
    subtitle_family = subtitle_family, subtitle_size = subtitle_size,
    subtitle_face = subtitle_face, subtitle_margin = subtitle_margin,
    axis_text_family = axis_text_family, axis_text_size = axis_text_size,
    axis_title_family = axis_title_family, axis_title_size = axis_title_size,
    axis_title_face = axis_title_face, axis_title_margin = axis_title_margin,
    axis_title_just = axis_title_just,
    caption_family = caption_family, caption_size = caption_size,
    caption_face = caption_face, caption_margin = caption_margin,
    strip_text_family = strip_text_family, strip_text_size = strip_text_size,
    strip_text_face = strip_text_face,
    plot_margin = plot_margin,
    grid_col = grid_col, grid = grid,
    axis_col = axis_col, axis = axis
    )

}


#' Import Recursive font for use in charts
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @md
#' @note This will take care of ensuring PDF/PostScript usage. The location of the
#'   font directory is displayed after the base import is complete. It is highly
#'   recommended that you install them on your system the same way you would any
#'   other font you wish to use in other programs.
#' @export
import_recursive <- function() {

  rec_font_dir <- system.file("fonts", "recursive", package="ewenthemes")

  suppressWarnings(suppressMessages(extrafont::font_import(rec_font_dir, prompt=FALSE)))

  message(
    sprintf(
      "You will likely need to install these fonts on your system as well.\n\nYou can find them in [%s]",
      rec_font_dir
      )
  )

}

#' @rdname Recursive
#' @md
#' @title Recursive font name R variable aliases
#' @description `font_rs` == "`Recursive`"
#' @format length 1 character vector
#' @export
font_rs <- "Recursive"

#' @rdname Recursive
#' @md
#' @note `font_rs_light` (a.k.a. "`Recursive Light`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_ws_light` == "`Recursive Light`"
#' @export
font_rs_light <- "Recursive Light"

#' @rdname Recursive
#' @md
#' @note `font_rs_extra_light` (a.k.a. "`Recursive ExtraLight`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_ws_extra_light` == "`Recursive ExtraLight`"
#' @export
font_rs_extra_light <- "Recursive ExtraLight"



R/work-sans.R
#' [ggplot2] theme supporting the \href{https://ewen.io/}{ewen.io} site, based on
#' `theme_ipsum` from \href{https://github.com/hrbrmstr/hrbrthemes}{hrbrthemes}.
#'
#' You should [import_work_sans]() first and also install the fonts on your
#' system before trying to use this theme.
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @inheritParams theme_ewen
#'
#' @export
#'
#' @examples \dontrun{
#' library(ggplot2)
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_ewen_ws()
#' }

theme_ewen_ws <- function(
  base_family = "Work Sans Light", base_colour = "#111111", base_size = 14,
  base_margin = 10, base_face = "plain",
  plot_title_family = "Work Sans", plot_title_size = 24,
  plot_title_face = "bold", plot_title_margin = 10,
  subtitle_family = base_family, subtitle_size = base_size,
  subtitle_face = base_face, subtitle_margin = base_margin,
  axis_text_family = base_family, axis_text_size = base_size,
  axis_title_family = base_family, axis_title_size = base_size,
  axis_title_face = base_face, axis_title_margin = base_margin,
  axis_title_just = "rt",
  caption_family = "Work Sans ExtraLight", caption_size = base_size,
  caption_face = base_face, caption_margin = base_margin*1.5,
  strip_text_family = base_family, strip_text_size = base_size,
  strip_text_face = base_face,
  plot_margin = c(10, 10, 10, 10),
  grid_col = "#cccccc", grid = "Y",
  axis_col = "#9B9B9B", axis = "x"
  ) {

  theme_ewen(
    base_family = base_family, base_colour = base_colour, base_size = base_size,
    base_margin = base_margin, base_face = base_face,
    plot_title_family = plot_title_family, plot_title_size = plot_title_size,
    plot_title_face = plot_title_face, plot_title_margin = plot_title_margin,
    subtitle_family = subtitle_family, subtitle_size = subtitle_size,
    subtitle_face = subtitle_face, subtitle_margin = subtitle_margin,
    axis_text_family = axis_text_family, axis_text_size = axis_text_size,
    axis_title_family = axis_title_family, axis_title_size = axis_title_size,
    axis_title_face = axis_title_face, axis_title_margin = axis_title_margin,
    axis_title_just = axis_title_just,
    caption_family = caption_family, caption_size = caption_size,
    caption_face = caption_face, caption_margin = caption_margin,
    strip_text_family = strip_text_family, strip_text_size = strip_text_size,
    strip_text_face = strip_text_face,
    plot_margin = plot_margin,
    grid_col = grid_col, grid = grid,
    axis_col = axis_col, axis = axis
    )

}


#' Import Work Sans font for use in charts
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @md
#' @note This will take care of ensuring PDF/PostScript usage. The location of the
#'   font directory is displayed after the base import is complete. It is highly
#'   recommended that you install them on your system the same way you would any
#'   other font you wish to use in other programs.
#' @export
import_work_sans <- function() {

  ws_font_dir <- system.file("fonts", "work-sans", package="ewenthemes")

  suppressWarnings(suppressMessages(extrafont::font_import(ws_font_dir, prompt=FALSE)))

  message(
    sprintf(
      "You will likely need to install these fonts on your system as well.\n\nYou can find them in [%s]",
      ws_font_dir)
  )

}

#' @rdname WorkSans
#' @md
#' @title Work Sans font name R variable aliases
#' @description `font_ws` == "`Work Sans`"
#' @format length 1 character vector
#' @export
font_ws <- "Work Sans"

#' @rdname WorkSans
#' @md
#' @note `font_ws_light` (a.k.a. "`Work Sans Light`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_ws_light` == "`Work Sans Light`"
#' @export
font_ws_light <- "Work Sans Light"

#' @rdname WorkSans
#' @md
#' @note `font_ws_extra_light` (a.k.a. "`Work Sans ExtraLight`") is not available on
#'     Windows and will throw a warning if used in plots.
#' @description `font_ws_extra_light` == "`Work Sans ExtraLight`"
#' @export
font_ws_extra_light <- "Work Sans ExtraLight"



R/theme-ewen.r
#' [ggplot2] theme supporting the \href{https://ewen.io/}{ewen.io} site, based on
#' `theme_ipsum` from \href{https://github.com/hrbrmstr/hrbrthemes}{hrbrthemes}.
#'
#' @section Building upon `theme_ewen`:
#' The function is setup in such a way that you can customize your own one by just
#' wrapping the call and changing the parameters. See source for examples.
#'
#' There is an option `ewenthemes.loadfonts` which -- if set to `TRUE` -- will
#' call `extrafont::loadfonts()` to register non-core fonts with R PDF & PostScript
#' devices. If you are running under Windows, the package calls the same function
#' to register non-core fonts with the Windows graphics device.
#'
#' @param base_family,base_colour,base_size,base_margin,base_face base font family, colour, size, margin and face
#' @param plot_title_family,plot_title_face,plot_title_size,plot_title_margin plot title family, face, size and margi
#' @param subtitle_family,subtitle_face,subtitle_size plot subtitle family, face and size
#' @param subtitle_margin plot subtitle margin bottom (single numeric value)
#' @param strip_text_family,strip_text_face,strip_text_size facet label font family, face and size
#' @param caption_family,caption_face,caption_size,caption_margin plot caption family, face, size and margin
#' @param axis_title_family,axis_title_face,axis_title_size,axis_title_margin axis title font family, face, size and margin
#' @param axis_title_just axis title font justification, one of `[blmcrt]`
#' @param plot_margin plot margin (specify with [ggplot2::margin()])
#' @param grid_col,axis_col grid & axis colors; grid defaults to `#cccccc`, axes default to `#2b2b2b`
#' @param grid panel grid (`TRUE`, `FALSE`, or a combination of `X`, `x`, `Y`, `y`)
#' @param axis_text_family,axis_text_size axis text font family and size
#' @param axis add x or y axes? `TRUE`, `FALSE`, "`xy`"
#'
#' @export
#'
#' @examples \dontrun{
#' library(ggplot2)
#'
#' ggplot(mtcars, aes(mpg, wt)) +
#'   geom_point() +
#'   labs(x="Fuel efficiency (mpg)", y="Weight (tons)",
#'        title="Seminal ggplot2 scatterplot example",
#'        subtitle="A plot that is only useful for demonstration purposes",
#'        caption="Brought to you by the letter 'g'") +
#'   theme_ewen()
#' }

theme_ewen <- function(
  base_family = "Arial Narrow", base_colour = "#111111", base_size = 14,
  base_margin = 10, base_face = "plain",
  plot_title_family = "Arial Narrow", plot_title_size = 24,
  plot_title_face = "bold", plot_title_margin = 10,
  subtitle_family = base_family, subtitle_size = base_size,
  subtitle_face = base_face, subtitle_margin = base_margin,
  axis_text_family = base_family, axis_text_size = base_size,
  axis_title_family = base_family, axis_title_size = base_size,
  axis_title_face = base_face, axis_title_margin = base_margin,
  axis_title_just = "rt",
  caption_family = base_family, caption_size = base_size,
  caption_face = base_face, caption_margin = base_margin*1.5,
  strip_text_family = base_family, strip_text_size = base_size,
  strip_text_face = base_face,
  plot_margin = c(10, 10, 10, 10),
  grid_col = "#cccccc", grid = "Y",
  axis_col = "#9B9B9B", axis = "x"
  ) {

  ret <- theme_minimal(base_family=base_family, base_size=base_size)

  ret <- ret + theme(
    legend.background = element_blank(), legend.key = element_blank()
    )

  if (inherits(grid, "character") | grid == TRUE) {

    ret <- ret + theme(
      panel.grid = element_line(colour = grid_col, size = 0.2),
      panel.grid.major = element_line(colour = grid_col, size = 0.2),
      panel.grid.minor = element_line(colour = grid_col, size = 0.15)
      )

    if (inherits(grid, "character")) {
      if (regexpr("X", grid)[1] < 0) ret <- ret + theme(panel.grid.major.x = element_blank())
      if (regexpr("Y", grid)[1] < 0) ret <- ret + theme(panel.grid.major.y = element_blank())
      if (regexpr("x", grid)[1] < 0) ret <- ret + theme(panel.grid.minor.x = element_blank())
      if (regexpr("y", grid)[1] < 0) ret <- ret + theme(panel.grid.minor.y = element_blank())
    }

  } else {
    ret <- ret + theme(panel.grid=element_blank())
  }

  if (inherits(axis, "character") | axis == TRUE) {
    ret <- ret + theme(axis.line = element_line(colour = "#2b2b2b", size = 0.7))
    if (inherits(axis, "character")) {
      axis <- tolower(axis)
      if (regexpr("x", axis)[1] < 0) {
        ret <- ret + theme(axis.line.x = element_blank())
      } else {
        ret <- ret + theme(axis.line.x = element_line(colour = axis_col, size = 0.7))
      }
      if (regexpr("y", axis)[1] < 0) {
        ret <- ret + theme(axis.line.y = element_blank())
      } else {
        ret <- ret + theme(axis.line.y = element_line(colour = axis_col, size = 0.7))
      }
    } else {
      ret <- ret + theme(axis.line = element_line(colour = axis_col, size = 0.7))
    }
  } else {
    ret <- ret + theme(axis.line = element_blank())
  }

    ret <- ret + theme(
      axis.ticks = element_blank(), axis.ticks.x = element_blank(),
      axis.ticks.y = element_blank()
      )

  xj <- switch(tolower(substr(axis_title_just, 1, 1)), b=0, l=0, m=0.5, c=0.5, r=1, t=1)
  yj <- switch(tolower(substr(axis_title_just, 2, 2)), b=0, l=0, m=0.5, c=0.5, r=1, t=1)

  ret <- ret + theme(

    plot.title = element_text(
      hjust=0, size=plot_title_size, margin=margin(b=plot_title_margin),
      family=plot_title_family, face=plot_title_face, colour = base_colour
      ),
    plot.title.position = "plot",
    plot.subtitle = element_text(
      hjust=0, size=subtitle_size, margin=margin(b=subtitle_margin),
      family=subtitle_family, face=subtitle_face, colour = base_colour
      ),

    axis.title = element_text(
      size = axis_title_size, family=axis_title_family, colour = base_colour,
      ),
    axis.title.x=element_text(
      hjust=xj, size=axis_title_size, family=axis_title_family, face=axis_title_face,
      margin = margin(t=axis_title_margin)
      ),
    axis.title.y=element_text(
      hjust=yj, size=axis_title_size, family=axis_title_family, face=axis_title_face,
      margin = margin(r=axis_title_margin)
      ),
    axis.title.y.right=element_text(
      hjust=yj, size=axis_title_size, angle=90, family=axis_title_family,
      face=axis_title_face, margin = margin(l=axis_title_margin)
      ),
    axis.text.x=element_text(
      size = axis_text_size, margin = margin(t=2.5), family = axis_text_family
      ),
    axis.text.y=element_text(
      size = axis_text_size, margin = margin(r=2.5), family = axis_text_family
      ),

    plot.caption=element_text(
      hjust=0, size=caption_size, margin=margin(t=caption_margin),
      family=caption_family, face=caption_face
      ),
    plot.caption.position = "plot",

    strip.text=element_text(
      hjust=0, size=strip_text_size, face=strip_text_face, family=strip_text_family
      ),

    panel.spacing=grid::unit(2, "lines"), plot.margin=margin(plot_margin)

    )

  ret

}

#' Update matching font defaults for text geoms
#'
#' Updates [ggplot2::geom_label] and [ggplot2::geom_text] font defaults
#'
#' @param family,face,size,color font family name, face, size and color
#' @export
update_geom_font_defaults <- function(family="Arial Narrow", face="plain", size=3.5,
                                      color = "#111111") {
  update_geom_defaults("text", list(family=family, face=face, size=size, color=color))
  update_geom_defaults("label", list(family=family, face=face, size=size, color=color))
}

#' @rdname ArialNarrow
#' @md
#' @title Arial Narrow font name R variable aliases
#' @description `font_an` == "`Arial Narrow`"
#' @format length 1 character vector
#' @export
font_an <- "Arial Narrow"
